firedrake/projection.py

```diff
@@ -83,21 +94,45 @@ def project(v, V, bcs=None, mesh=None,
         raise RuntimeError('Shape mismatch between source %s and target function spaces %s in project' %
                            (v.ufl_shape, ret.ufl_shape))
 
-    p = ufl_expr.TestFunction(V)
-    q = ufl_expr.TrialFunction(V)
-    a = ufl.inner(p, q) * ufl.dx(domain=V.mesh())
-    L = ufl.inner(p, v) * ufl.dx(domain=V.mesh())
+    if method == "l2":
+        # Perform standard L2 projection
+        p = ufl_expr.TestFunction(V)
+        q = ufl_expr.TrialFunction(V)
+        a = ufl.inner(p, q) * ufl.dx(domain=V.mesh())
+        L = ufl.inner(p, v) * ufl.dx(domain=V.mesh())
+
+        # Default to 1e-8 relative tolerance
+        if solver_parameters is None:
+            solver_parameters = {'ksp_type': 'cg', 'ksp_rtol': 1e-8}
+        else:
+            solver_parameters.setdefault('ksp_type', 'cg')
+            solver_parameters.setdefault('ksp_rtol', 1e-8)
+
+        _solve(a == L, ret, bcs=bcs,
+               solver_parameters=solver_parameters,
+               form_compiler_parameters=form_compiler_parameters)
+
+    elif method == "average":
+        # Loop over node extent and dof extent
+        shapes = (V.finat_element.space_dimension(), np.prod(V.shape))
+        accumulate_kernel = """
+        for (int i=0; i<%d; ++i) {
+            for (int j=0; j<%d; ++j) {
+                vo[i][j] += v[i][j];
+                w[i][j] += 1.0;
+        }}""" % shapes
+
+        # Ensure function we populate into is zeroed out
+        ret.assign(0.0)
+        w = function.Function(V)
+        par_loop(accumulate_kernel, ufl.dx, {"vo": (ret, INC),
+                                             "w": (w, INC),
+                                             "v": (v, READ)})
+        ret.dat /= w.dat
 
-    # Default to 1e-8 relative tolerance
-    if solver_parameters is None:
-        solver_parameters = {'ksp_type': 'cg', 'ksp_rtol': 1e-8}
     else:
-        solver_parameters.setdefault('ksp_type', 'cg')
-        solver_parameters.setdefault('ksp_rtol', 1e-8)
+        raise ValueError("Method type %s not recognized" % str(method))
 
-    _solve(a == L, ret, bcs=bcs,
-           solver_parameters=solver_parameters,
-           form_compiler_parameters=form_compiler_parameters)
     return ret
```

Comment:
This feels like good opportunity to reduce code duplication by making `project` do something like:

```
return Projector(...).project()
```